README.md
# push-all-in-one-action

基于 [push-all-in-one](https://github.com/CaoMeiYouRen/push-all-in-one) 的定时推送 action，支持 Server酱、酷推、邮件、钉钉机器人、企业微信机器人、企业微信应用、pushplus、iGot 等多种推送方式。

本项目也欢迎二次开发。

## 使用

### 方法一：fork 本项目

点击右上角的 `fork` 按钮，然后进入 `Actions` 页面开启`Workflows` 下的  `Send`，然后请阅读[配置](#配置)一栏

>   此处借用了 `BiliExp` 的图片，配置过程是一样的

![](https://cdn.jsdelivr.net/gh/CaoMeiYouRen/image-hosting-01@master/images/20210304232639.png)

### 方法二：配置 GitHub workflows

在项目根目录下新建 `.github/workflows/send.yml` 然后复制如下内容

```yaml
name: "Send"

on:
  schedule:
    - cron: "0 23 * * *" # UTC+8 的 7 点执行，修改时注意时区
  watch:
    types: started
  workflow_dispatch:

jobs:
  send:
    name: "Push All In One"
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - name: "执行 Push All In One"
        uses: CaoMeiYouRen/push-all-in-one-action@master
        env:
          SEND_TITLE: ${{secrets.SEND_TITLE}}
          SEND_DESP: ${{secrets.SEND_DESP}}
          SCTKEY: ${{secrets.SCTKEY}}
          COOL_PUSH_SKEY: ${{secrets.COOL_PUSH_SKEY}}
          COOL_PUSH_TYPE: ${{secrets.COOL_PUSH_TYPE}}
          BER_KEY: ${{secrets.BER_KEY}}
          EMAIL_ADDRESS: ${{secrets.EMAIL_ADDRESS}}
          DINGTALK_ACCESS_TOKEN: ${{secrets.DINGTALK_ACCESS_TOKEN}}
          DINGTALK_SECRET: ${{secrets.DINGTALK_SECRET}}
          WX_ROBOT_KEY: ${{secrets.WX_ROBOT_KEY}}
          WX_ROBOT_MSG_TYPE: ${{secrets.WX_ROBOT_MSG_TYPE}}
          WX_APP_CORPID: ${{secrets.WX_APP_CORPID}}
          WX_APP_AGENTID: ${{secrets.WX_APP_AGENTID}}
          WX_APP_SECRET: ${{secrets.WX_APP_SECRET}}
          WX_APP_USERID: ${{secrets.WX_APP_USERID}}
          PUSH_PLUS_TOKEN: ${{secrets.PUSH_PLUS_TOKEN}}
          PUSH_PLUS_TEMPLATE_TYPE: ${{secrets.PUSH_PLUS_TEMPLATE_TYPE}}
          I_GOT_KEY: ${{secrets.I_GOT_KEY}}

```

## 配置

![image](https://cdn.jsdelivr.net/gh/CaoMeiYouRen/image-hosting-01@master/images/20210304232805.png)

>   此处依旧借用了 `BiliExp` 的图片，注意下 `Name` 和 `Value` 别填成上面的内容就行了，按照下面列表的说明填写
>
>   在`Settings`-->`Secrets`里添加的参数，`Name`必须为下列的参数名称之一，`Value`则填写对应获取的值

本项目所有的配置都采用 `Actions Secrets` 的方式来配置，以下是配置说明。

没有配置的推送方式不会执行推送的。

| secrets                 | 说明                                                         |
| ----------------------- | ------------------------------------------------------------ |
| SEND_TITLE              | 要发送的标题，必填                                           |
| SEND_DESP               | 要发送的内容，选填                                           |
| SCTKEY                  | Server酱·Turbo  SCTKEY。官方文档：https://sct.ftqq.com/      |
| COOL_PUSH_SKEY          | 酷推 SKEY。官方文档：https://cp.xuthus.cc/                   |
| COOL_PUSH_TYPE          | 推送类型，默认为` send`                                      |
| BER_KEY                 | BER分邮件系统。官方文档：http://doc.berfen.com/1239397       |
| EMAIL_ADDRESS           | 推送到的邮箱                                                 |
| DINGTALK_ACCESS_TOKEN   | 钉钉机器人 access_token。官方文档：https://developers.dingtalk.com/document/app/custom-robot-access |
| DINGTALK_SECRET         | 钉钉机器人加签安全秘钥（HmacSHA256）                         |
| WX_ROBOT_KEY            | 企业微信群机器人。官方文档：https://work.weixin.qq.com/help?person_id=1&doc_id=13376 |
| WX_ROBOT_MSG_TYPE       | 消息类型，默认 `text`                                        |
| WX_APP_CORPID           | 企业微信企业ID，获取方式参考 https://work.weixin.qq.com/api/doc/90000/90135/91039#14953/corpid |
| WX_APP_AGENTID          | 企业应用的id。企业内部开发，可在应用的设置页面查看 |
| WX_APP_SECRET           | 应用的凭证密钥，获取方式参考：https://work.weixin.qq.com/api/doc/90000/90135/91039#14953/secret |
| WX_APP_USERID           | 指定接收消息的成员。若不指定则默认为 ”@all”。 |
| PUSH_PLUS_TOKEN                   | pushplus 推送加开放平台。官方文档：http://pushplus.hxtrip.com/doc/ |
| PUSH_PLUS_TEMPLATE_TYPE | 发送消息模板，默认为 html |
| I_GOT_KEY               | iGot 推送，官方文档：https://wahao.github.io/Bark-MP-helper |


CHANGELOG.md
# push-all-in-one-action

## [1.0.4](https://github.com/CaoMeiYouRen/push-all-in-one-action/compare/v1.0.3...v1.0.4) (2021-12-24)


### 🐛 Bug 修复

* 修改 CI 配置 ([91a49ee](https://github.com/CaoMeiYouRen/push-all-in-one-action/commit/91a49ee))
* 更新 依赖 ([cb75c22](https://github.com/CaoMeiYouRen/push-all-in-one-action/commit/cb75c22))


* Merge branch 'master' of github.com:CaoMeiYouRen/push-all-in-one-action ([4778407](https://github.com/CaoMeiYouRen/push-all-in-one-action/commit/4778407))

## [1.0.3](https://github.com/CaoMeiYouRen/push-all-in-one-action/compare/v1.0.2...v1.0.3) (2021-03-20)


### 🐛 Bug Fixes

* 修改代码组织 ([b4f45ae](https://github.com/CaoMeiYouRen/push-all-in-one-action/commit/b4f45ae))

## [1.0.2](https://github.com/CaoMeiYouRen/push-all-in-one-action/compare/v1.0.1...v1.0.2) (2021-03-05)


### 🐛 Bug Fixes

* 修改提示信息；修改 test.yml ([f410840](https://github.com/CaoMeiYouRen/push-all-in-one-action/commit/f410840))

## [1.0.1](https://github.com/CaoMeiYouRen/push-all-in-one-action/compare/v1.0.0...v1.0.1) (2021-03-04)


### 🐛 Bug Fixes

* 修改 ts target 为 es2019 ([e10db8d](https://github.com/CaoMeiYouRen/push-all-in-one-action/commit/e10db8d))

# 1.0.0 (2021-03-04)


### ✨ Features

* 完成功能开发和文档编写 ([b415ccd](https://github.com/CaoMeiYouRen/push-all-in-one-action/commit/b415ccd))


__tests__/main.test.ts
import { wait } from '../src/wait'
import * as process from 'process'
import * as cp from 'child_process'
import * as path from 'path'

test('throws invalid number', async () => {
    const input = parseInt('foo', 10)
    await expect(wait(input)).rejects.toThrow('milliseconds not a number')
})

test('wait 500 ms', async () => {
    const start = new Date()
    await wait(500)
    const end = new Date()
    const delta = Math.abs(end.getTime() - start.getTime())
    expect(delta).toBeGreaterThan(450)
})

// shows how the runner will run a javascript action with env / stdout protocol
test('test runs', () => {
    process.env['INPUT_MILLISECONDS'] = '500'
    const np = process.execPath
    const ip = path.join(__dirname, '..', 'lib', 'main.js')
    const options: cp.ExecFileSyncOptions = {
        env: process.env,
    }
    console.log(cp.execFileSync(np, [ip], options).toString())
})


src/push.ts
import { AxiosResponse } from 'axios'
import { ServerChanTurbo, CoolPush, Dingtalk, Email, WechatRobot, WechatApp, PushPlus, IGot } from 'push-all-in-one'
import {
    SCTKEY, COOL_PUSH_SKEY, COOL_PUSH_TYPE, BER_KEY, EMAIL_ADDRESS, DINGTALK_ACCESS_TOKEN, DINGTALK_SECRET,
    WX_ROBOT_KEY, WX_APP_CORPID, WX_APP_AGENTID, WX_APP_SECRET, WX_APP_USERID, PUSH_PLUS_TOKEN, I_GOT_KEY, PUSH_PLUS_TEMPLATE_TYPE,
    WX_ROBOT_MSG_TYPE,
} from './env'
import { info, warn } from './help'

export async function runPushAllInOne(title: string, desp?: string): Promise<PromiseSettledResult<AxiosResponse<any>>[]> {
    const pushs: Promise<AxiosResponse<any>>[] = []
    if (SCTKEY) {
        // Server酱。官方文档：https://sct.ftqq.com/
        const serverChanTurbo = new ServerChanTurbo(SCTKEY)
        pushs.push(serverChanTurbo.send(title, desp))
        info('Server酱·Turbo 已加入推送队列')
    } else {
        info('未配置 Server酱·Turbo，已跳过')
    }

    if (COOL_PUSH_SKEY) {
        // 酷推。官方文档：https://cp.xuthus.cc/
        const coolPush = new CoolPush(COOL_PUSH_SKEY)
        pushs.push(coolPush.send(`${title}\n${desp}`, COOL_PUSH_TYPE))
        info('Cool Push 已加入推送队列')
    } else {
        info('未配置 Cool Push，已跳过')
    }

    if (EMAIL_ADDRESS) {
        // BER分邮件系统。官方文档：http://doc.berfen.com/1239397
        // 如果不提供 BER_KEY 将会使用免费版本进行推送。免费接口有较多限制，请自行斟酌
        const email = new Email(BER_KEY)
        pushs.push(email.send({
            title,
            // subtitle: '这是个小标题',
            desp,
            address: EMAIL_ADDRESS,
        }))
        info('BER分邮件系统 已加入推送队列')
    } else {
        info('未配置 BER分邮件系统，已跳过')
    }

    if (DINGTALK_ACCESS_TOKEN) {
        // 钉钉机器人。官方文档：https://developers.dingtalk.com/document/app/custom-robot-access
        const dingtalk = new Dingtalk(DINGTALK_ACCESS_TOKEN, DINGTALK_SECRET)
        pushs.push(dingtalk.send(title, desp))
        info('钉钉机器人 已加入推送队列')
    } else {
        info('未配置 钉钉机器人，已跳过')
    }

    if (WX_ROBOT_KEY) {
        // 企业微信群机器人。官方文档：https://work.weixin.qq.com/help?person_id=1&doc_id=13376
        // 企业微信群机器人的使用需要两人以上加入企业，如果个人使用微信推送建议使用 企业微信应用+微信插件 推送
        const wechatRobot = new WechatRobot(WX_ROBOT_KEY)
        pushs.push(wechatRobot.send(`${title}\n${desp}`, WX_ROBOT_MSG_TYPE))
        info('企业微信群机器人 已加入推送队列')
    } else {
        info('未配置 企业微信群机器人，已跳过')
    }

    if (WX_APP_CORPID && WX_APP_AGENTID && WX_APP_SECRET) {
        // 企业微信应用推送，官方文档：https://work.weixin.qq.com/api/doc/90000/90135/90664
        const wechatApp = new WechatApp({
            WX_APP_CORPID,
            WX_APP_AGENTID,
            WX_APP_SECRET,
            WX_APP_USERID,
        })
        pushs.push(wechatApp.send(`${title}\n${desp}`))
        info('企业微信应用推送 已加入推送队列')
    } else {
        info('未配置 企业微信应用推送，已跳过')
    }

    if (PUSH_PLUS_TOKEN) {
        // pushplus 推送，官方文档：http://pushplus.hxtrip.com/doc/
        const pushplus = new PushPlus(PUSH_PLUS_TOKEN)
        pushs.push(pushplus.send(title, desp, PUSH_PLUS_TEMPLATE_TYPE))
        info('pushplus推送 已加入推送队列')
    } else {
        info('未配置 pushplus推送，已跳过')
    }

    if (I_GOT_KEY) {
        // iGot 推送，官方文档：https://wahao.github.io/Bark-MP-helper
        const iGot = new IGot(I_GOT_KEY)
        pushs.push(iGot.send(title, desp))
        info('iGot推送 已加入推送队列')
    } else {
        info('未配置 iGot推送，已跳过')
    }
    if (pushs.length === 0) {
        warn('未配置任何推送，请检查推送配置的环境变量！')
        return []
    }

    const results = await Promise.allSettled(pushs)
    const success = results.filter((e) => e.status === 'fulfilled')
    const fail = results.filter((e) => e.status === 'rejected')

    info(`本次共推送 ${results.length} 个，成功 ${success.length} 个，失败 ${fail.length} 个`)

    return results
}


.releaserc.js
const { name } = require('./package.json')
module.exports = {
    plugins: [
        [
            "@semantic-release/commit-analyzer",
            {
                "config": "conventional-changelog-cmyr-config"
            }
        ],
        ["@semantic-release/release-notes-generator",
            {
                "config": "conventional-changelog-cmyr-config"
            }],
        [
            "@semantic-release/changelog",
            {
                "changelogFile": "CHANGELOG.md",
                "changelogTitle": "# " + name
            }
        ],
        '@semantic-release/npm',
        '@semantic-release/github',
        [
            "@semantic-release/git",
            {
                "assets": [
                    "dist",
                    "src",
                    "CHANGELOG.md",
                    "package.json"
                ]
            }
        ]
    ]
}


rollup.config.js
import path from 'path'
import { nodeResolve } from '@rollup/plugin-node-resolve'
import { terser } from 'rollup-plugin-terser'
import commonjs from '@rollup/plugin-commonjs'
import typescript from '@rollup/plugin-typescript'
import json from '@rollup/plugin-json'
import analyzer from 'rollup-plugin-analyzer'
import alias from '@rollup/plugin-alias'
import _ from 'lodash'
import { dependencies, name } from './package.json'
const external = Object.keys(dependencies) // 默认不打包 dependencies
const outputName = _.upperFirst(_.camelCase(name))// 导出的模块名称 PascalCase
const env = process.env
// const __PROD__ = env.NODE_ENV === 'production'
// const __DEV__ = env.NODE_ENV === 'development'
const __ANALYZER__ = Boolean(env.ANALYZER)
function getPlugins({ isBrowser = false, isMin = false, isDeclaration = false }) {
    const plugins = []
    plugins.push(
        nodeResolve({
            browser: isBrowser,
            preferBuiltins: true,
        }),
    )
    plugins.push(
        typescript({
            tsconfig: 'tsconfig.json',
            module: 'esnext',
            target: 'es2019',
            esModuleInterop: true,
            allowSyntheticDefaultImports: true,
            declaration: isDeclaration,
        }),
    )
    plugins.push(alias({
        entries: [
            { find: '@', replacement: path.resolve(__dirname, '../src') },
        ],
    }))
    plugins.push(
        commonjs({
            sourceMap: false,
        }),
    )
    plugins.push(
        json({}),
    )
    if (isMin) {
        plugins.push(
            terser({
                module: true,
            }),
        )
    }
    if (__ANALYZER__) {
        plugins.push(
            analyzer({
            }),
        )
    }
    return plugins
}

export default [
    {
        input: 'src/index.ts',
        // external,
        output: {
            file: 'dist/index.js', // 生成 cjs
            format: 'cjs',
            name: outputName,
            sourcemap: true,
        },
        plugins: getPlugins({
            isBrowser: false,
            isDeclaration: false,
            isMin: false,
        }),
    },
    // {
    //     input: 'src/index.ts',
    //     external,
    //     output: {
    //         file: 'lib/index.esm.js',
    //         format: 'esm',
    //         name: outputName,
    //         sourcemap: true,
    //     },
    //     plugins: getPlugins({
    //         isBrowser: false,
    //         isDeclaration: false,
    //         isMin: false,
    //     }),
    // },
]


src/env.ts
import { MsgType, PushType, TemplateType } from 'push-all-in-one'

const env = process.env
/**
 * 要发送的标题，必填
 */
export const SEND_TITLE = env.SEND_TITLE || ''
/**
 * 要发送的内容，选填
 */
export const SEND_DESP = env.SEND_DESP || ''
/**
 * Server酱。官方文档：https://sct.ftqq.com/
 */
export const SCTKEY = env.SCTKEY || ''
/**
 * 酷推。官方文档：https://cp.xuthus.cc/
 */
export const COOL_PUSH_SKEY = env.COOL_PUSH_SKEY || ''
/**
 * 推送类型 'send' | 'group' | 'psend' | 'pgroup' | 'wx' | 'tg'
 */
export const COOL_PUSH_TYPE = (env.COOL_PUSH_TYPE || 'send') as PushType

/**
 * BER分邮件系统。官方文档：http://doc.berfen.com/1239397
 */
export const BER_KEY = env.BER_KEY || ''
/**
 * 推送到的邮箱
 */
export const EMAIL_ADDRESS = env.EMAIL_ADDRESS || ''
/**
 * 钉钉机器人。官方文档：https://developers.dingtalk.com/document/app/custom-robot-access
 */
export const DINGTALK_ACCESS_TOKEN = env.DINGTALK_ACCESS_TOKEN || ''
/**
 * 钉钉机器人加签安全秘钥（HmacSHA256）
 */
export const DINGTALK_SECRET = env.DINGTALK_SECRET || ''
/**
 * 企业微信群机器人。官方文档：https://work.weixin.qq.com/help?person_id=1&doc_id=13376
 */
export const WX_ROBOT_KEY = env.WX_ROBOT_KEY || ''
/**
 * 消息类型  'text' | 'markdown'
 */
export const WX_ROBOT_MSG_TYPE = (env.WX_ROBOT_MSG_TYPE || 'text') as MsgType

/**
 * 企业微信应用推送，官方文档：https://work.weixin.qq.com/api/doc/90000/90135/90664
 */
export const WX_APP_CORPID = env.WX_APP_CORPID || ''

export const WX_APP_AGENTID = Number(env.WX_APP_AGENTID)

export const WX_APP_SECRET = env.WX_APP_SECRET || ''

export const WX_APP_USERID = env.WX_APP_USERID || ''
/**
 * pushplus 推送，官方文档：http://pushplus.hxtrip.com/doc/
 */
export const PUSH_PLUS_TOKEN = env.PUSH_PLUS_TOKEN || ''
// 消息模板类型 'html' | 'json' | 'cloudMonitor'
export const PUSH_PLUS_TEMPLATE_TYPE = (env.PUSH_PLUS_TEMPLATE_TYPE || 'html') as TemplateType
/**
 * iGot 推送，官方文档：https://wahao.github.io/Bark-MP-helper
 */
export const I_GOT_KEY = env.I_GOT_KEY || ''
